pkg/authorization/controller/authorizationsync/origin_to_rbac_test.go

```diff
@@ -0,0 +1,545 @@
+	// ought to have one action writing status
+	if len(actions) != 1 {
+		return nil, fmt.Errorf("expected %v, got %v", 1, actions)
+	}
+
+	action, ok := actions[0].(clienttesting.DeleteAction)
+	if !ok {
+		return nil, fmt.Errorf("expected %v, got %v", "update", actions[0])
+	}
+
+	return action, nil
+}
+
+func TestSyncClusterRole(t *testing.T) {
+	tests := []struct {
+		name string
+
+		key            string
+		startingRBAC   []*rbac.ClusterRole
+		startingOrigin []*authorizationapi.ClusterRole
+
+		actionCheck func([]clienttesting.Action) error
+	}{
+		{
+			name: "no action on missing both",
+			key:  "resource-01",
+			actionCheck: func(actions []clienttesting.Action) error {
+				if len(actions) != 0 {
+					return fmt.Errorf("expected %v, got %v", 0, actions)
+				}
+				return nil
+			},
+		},
+		{
+			name: "simple create",
+			key:  "resource-01",
+			startingOrigin: []*authorizationapi.ClusterRole{
+				{ObjectMeta: metav1.ObjectMeta{Name: "resource-01"}},
+			},
+			actionCheck: func(actions []clienttesting.Action) error {
+				action, err := ensureSingleCreateAction(actions)
+				if err != nil {
+					return err
+				}
+				if e, a := "resource-01", action.GetObject().(*rbac.ClusterRole).Name; e != a {
+					return fmt.Errorf("expected %v, got %v", e, a)
+				}
+				return nil
+			},
+		},
+		{
+			name: "delete on missing origin",
+			key:  "resource-01",
+			startingRBAC: []*rbac.ClusterRole{
+				{ObjectMeta: metav1.ObjectMeta{Name: "resource-01"}},
+			},
+			actionCheck: func(actions []clienttesting.Action) error {
+				action, err := ensureSingleDeleteAction(actions)
+				if err != nil {
+					return err
+				}
+				if e, a := "resource-01", action.GetName(); e != a {
+					return fmt.Errorf("expected %v, got %v", e, a)
+				}
+				return nil
+			},
+		},
+		{
+			name: "simple update",
+			key:  "resource-01",
+			startingRBAC: []*rbac.ClusterRole{
+				{ObjectMeta: metav1.ObjectMeta{Name: "resource-01"}},
+			},
+			startingOrigin: []*authorizationapi.ClusterRole{
+				{ObjectMeta: metav1.ObjectMeta{Name: "resource-01", Annotations: map[string]string{"foo": "different"}}},
+			},
+			actionCheck: func(actions []clienttesting.Action) error {
+				action, err := ensureSingleUpdateAction(actions)
+				if err != nil {
+					return err
+				}
+				if e, a := "resource-01", action.GetObject().(*rbac.ClusterRole).Name; e != a {
+					return fmt.Errorf("expected %v, got %v", e, a)
+				}
+				return nil
+			},
+		},
+		{
+			name: "no action on zero diff",
+			key:  "resource-01",
+			startingRBAC: []*rbac.ClusterRole{
+				{ObjectMeta: metav1.ObjectMeta{Name: "resource-01"}},
+			},
+			startingOrigin: []*authorizationapi.ClusterRole{
+				{ObjectMeta: metav1.ObjectMeta{Name: "resource-01"}},
+			},
+			actionCheck: func(actions []clienttesting.Action) error {
+				if len(actions) != 0 {
+					return fmt.Errorf("expected %v, got %v", 0, actions)
+				}
+				return nil
+			},
+		},
```

Comment:
This applies to other controllers as well (except for the annotation, which is only for `ClusterRoles`).